Hive and Spark EC have a ClassNotFoundException bug when using jar UDF

**Describe the bug**
Hive and Spark EC have a ClassNotFoundException bug when using jar UDF. error info:
```
Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: *.CardNoFirstEightMask
        at org.apache.hadoop.hive.ql.udf.generic.GenericUDFBridge.getUdfClass(GenericUDFBridge.java:134)
```

**To Reproduce**
Steps to reproduce the behavior:
1. create udf
2. use udf by hql or sql
3.  See error

--- a/linkis-computation-governance/linkis-engineconn-manager/linkis-engineconn-manager-server/src/main/scala/org/apache/linkis/ecm/server/hook/JarUDFLoadECMHook.scala
+++ b/linkis-computation-governance/linkis-engineconn-manager/linkis-engineconn-manager-server/src/main/scala/org/apache/linkis/ecm/server/hook/JarUDFLoadECMHook.scala
@@ -31,7 +31,7 @@ class JarUDFLoadECMHook extends ECMHook with Logging {
         info("start loading UDFs")
         val udfInfos = UDFClient.getUdfInfos(request.user,"udf").filter{ info => info.getUdfType == 0 && info.getExpire == false && StringUtils.isNotBlank(info.getPath) && info.getLoad == true }
         udfInfos.foreach{ udfInfo =>
-          LaunchConstants.addPathToClassPath(pel.environment, udfInfo.getPath)
+          LaunchConstants.addPathToUDFPath(pel.environment, udfInfo.getPath)
         }
     }
   }

--- a/linkis-engineconn-plugins/engineconn-plugins/hive/src/main/resources/linkis-engineconn.properties
+++ b/linkis-engineconn-plugins/engineconn-plugins/hive/src/main/resources/linkis-engineconn.properties
@@ -19,4 +19,4 @@ wds.linkis.engineconn.debug.enable=true
 #wds.linkis.keytab.enable=true
 wds.linkis.engineconn.plugin.default.class=org.apache.linkis.engineplugin.hive.HiveEngineConnPlugin
 wds.linkis.bdp.hive.init.sql.enable=true
-wds.linkis.engine.connector.hooks=org.apache.linkis.engineconn.computation.executor.hook.ComputationEngineConnHook,org.apache.linkis.engineconn.computation.executor.hook.JarUdfEngineHook
+wds.linkis.engine.connector.hooks=org.apache.linkis.engineconn.computation.executor.hook.ComputationEngineConnHook,org.apache.linkis.engineplugin.hive.hook.HiveAddJarsEngineHook,org.apache.linkis.engineconn.computation.executor.hook.JarUdfEngineHook

--- a/linkis-engineconn-plugins/engineconn-plugins/hive/src/main/scala/org/apache/linkis/engineplugin/hive/hook/HiveAddJarsEngineHook.scala
+++ b/linkis-engineconn-plugins/engineconn-plugins/hive/src/main/scala/org/apache/linkis/engineplugin/hive/hook/HiveAddJarsEngineHook.scala
@@ -27,6 +27,8 @@ import org.apache.linkis.engineplugin.hive.executor.HiveEngineConnExecutor
 import org.apache.linkis.manager.label.entity.Label
 import org.apache.linkis.manager.label.entity.engine.{CodeLanguageLabel, RunType}
 import org.apache.commons.lang.StringUtils
+import org.apache.linkis.common.conf.CommonVars
+import org.apache.linkis.manager.engineplugin.common.launch.process.Environment
 
 import scala.collection.JavaConversions._
 
@@ -43,9 +45,18 @@ class HiveAddJarsEngineHook extends EngineConnHook with Logging {
   override def afterExecutionExecute(engineCreationContext: EngineCreationContext, engineConn: EngineConn): Unit = Utils.tryAndError {
     val options = engineCreationContext.getOptions
     var jars: String = ""
+    val udf_jars = CommonVars(Environment.UDF_JARS.toString, "", "UDF jar PAth").getValue
+    logger.info("udf jar_path:" + udf_jars)
     options foreach {
       case (key, value) => if (JARS.equals(key)) {
         jars = value
+      }
+    }
+    if (StringUtils.isNotEmpty(udf_jars)) {
+      if (StringUtils.isNotEmpty(jars)) {
+        jars = jars + "," + udf_jars
+      } else {
+        jars = udf_jars
       }
     }
     val codeLanguageLabel = new CodeLanguageLabel

--- a/linkis-engineconn-plugins/engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/launch/SparkSubmitProcessEngineConnLaunchBuilder.scala
+++ b/linkis-engineconn-plugins/engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/launch/SparkSubmitProcessEngineConnLaunchBuilder.scala
@@ -33,16 +33,13 @@ import org.apache.linkis.manager.label.entity.Label
 import org.apache.linkis.manager.label.entity.engine.UserCreatorLabel
 import org.apache.linkis.protocol.UserWithCreator
 import org.apache.commons.lang.StringUtils
+
 import scala.collection.JavaConverters._
 import scala.collection.mutable.ArrayBuffer
 import org.apache.linkis.hadoop.common.conf.HadoopConf
 
 
-/**
 
- *
- * @Date 2020/10/23
- */
 class SparkSubmitProcessEngineConnLaunchBuilder private extends JavaProcessEngineConnLaunchBuilder {
 
   private[this] val fsRoot = "hdfs://"
@@ -172,9 +169,11 @@ class SparkSubmitProcessEngineConnLaunchBuilder private extends JavaProcessEngin
 
     //addOpt("--jars",Some(ENGINEMANAGER_JAR.getValue))
 //    info("No need to add jars for " + _jars.map(fromPath).exists(x => x.equals("hdfs:///")).toString())
-    _jars = _jars.filter(_.isNotBlankPath())
-
-    if(_jars.nonEmpty) {
+    if (_jars.isEmpty) {
+      _jars += AbsolutePath("")
+    }
+    _jars += AbsolutePath(variable(UDF_JARS))
+    if (_jars.nonEmpty) {
       addList("--jars", _jars.map(fromPath))
     }
 

--- a/linkis-engineconn-plugins/linkis-engineconn-plugin-framework/linkis-engineconn-plugin-core/src/main/scala/org/apache/linkis/manager/engineplugin/common/launch/process/Environment.scala
+++ b/linkis-engineconn-plugins/linkis-engineconn-plugin-framework/linkis-engineconn-plugin-core/src/main/scala/org/apache/linkis/manager/engineplugin/common/launch/process/Environment.scala
@@ -17,13 +17,15 @@
  
 package org.apache.linkis.manager.engineplugin.common.launch.process
 
+import org.apache.commons.lang.StringUtils
+
 
 object Environment extends Enumeration {
 
   type Environment = Value
   val USER, ECM_HOME, PWD, PATH, SHELL, JAVA_HOME, CLASSPATH,
       HADOOP_HOME, HADOOP_CONF_DIR, HIVE_CONF_DIR, LOG_DIRS, TEMP_DIRS,
-      ECM_HOST, ECM_PORT, RANDOM_PORT, SERVICE_DISCOVERY,EUREKA_PREFER_IP,ENGINECONN_ENVKEYS = Value
+      ECM_HOST, ECM_PORT, RANDOM_PORT, SERVICE_DISCOVERY,EUREKA_PREFER_IP, UDF_JARS, ENGINECONN_ENVKEYS = Value
 
   def variable(environment: Environment): String = LaunchConstants.EXPANSION_MARKER_LEFT + environment + LaunchConstants.EXPANSION_MARKER_RIGHT
 
@@ -44,6 +46,14 @@ object LaunchConstants {
       env.get(Environment.CLASSPATH.toString) + CLASS_PATH_SEPARATOR + value
     } else value
     env.put(Environment.CLASSPATH.toString, v)
+  }
+
+  def addPathToUDFPath(env: java.util.Map[String, String], value: String): Unit = {
+    if (StringUtils.isBlank(value)) return
+    val v = if(env.containsKey(Environment.UDF_JARS.toString)) {
+      env.get(Environment.UDF_JARS.toString) + "," + value
+    } else value
+    env.put(Environment.UDF_JARS.toString, v)
   }
 
 }

